kaldb/src/main/java/com/slack/kaldb/blobfs/s3/S3CrtBlobFs.java

```diff
@@ -0,0 +1,548 @@
+package com.slack.kaldb.blobfs.s3;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.slack.kaldb.blobfs.BlobFs;
+import com.slack.kaldb.blobfs.BlobFsConfig;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3CrtAsyncClientBuilder;
+import software.amazon.awssdk.services.s3.model.CopyObjectRequest;
+import software.amazon.awssdk.services.s3.model.CopyObjectResponse;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.DeleteObjectResponse;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+import software.amazon.awssdk.services.s3.model.MetadataDirective;
+import software.amazon.awssdk.services.s3.model.NoSuchKeyException;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+public class S3CrtBlobFs extends BlobFs {
+  public static final String S3_SCHEME = "s3://";
+  private static final Logger LOG = LoggerFactory.getLogger(S3CrtBlobFs.class);
+  private static final String DELIMITER = "/";
+
+  private final S3AsyncClient s3AsyncClient;
+
+  public S3CrtBlobFs(S3AsyncClient s3AsyncClient) {
+    this.s3AsyncClient = s3AsyncClient;
+  }
+
+  static boolean isNullOrEmpty(String target) {
+    return target == null || "".equals(target);
+  }
+
+  public static S3AsyncClient initS3Client(KaldbConfigs.S3Config config) {
```

Comment:
This is automatically called from the KalDB initialization code here -https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/server/Kaldb.java#L84. Since S3 is used widely across the application, we instantiate/initiate it once, and then pass this reference to all the classes that need it. This is somewhat similar to how we handle the metadata stores as well.